parla/task_runtime.py

```diff
@@ -1778,40 +1817,40 @@ def _schedule_tasks(self):
     # The mapper only maps tasks which have enough resources to run.
     # It's guaranteed for the task to have enough resources
 
-    def _launch_cpu_task(self, queue, task: Task, dev: Device):
+    def _launch_task(self, queue, dev: Device):
+        task = queue.pop()
         worker = self._free_worker_threads.pop() # grab a worker
-        logger.info(f"[Scheduler] Launching CPU task, %r on %r",
-                    task, worker)
-        # Assign the task to the worker (this notifies the worker's monitor)
-        worker.assign_task(task)
-        logger.debug(f"[Scheduler] Launched %r", task)
-
-    def _launch_gpu_task(self, queue, task: Task, dev: Device):
-        worker = self._free_worker_threads.pop() # grab a worker
-        logger.info(f"[Scheduler] Launching GPU task, %r on %r",
-                    task, worker)
-        # Assign the task to the worker (this notifies the worker's monitor)
+        logger.info(f"[Scheduler] Launching %s task, %r on %r",
+                    dev.architecture.id, task, worker)
         worker.assign_task(task)
         logger.debug(f"[Scheduler] Launched %r", task)
+        for dev in task.req.environment.placement:
+            self.update_launched_task_count(task, dev, 1)
 
     def _launch_tasks(self):
         """ Iterate through free devices and launch tasks on them
         """
         #logger.debug("[Scheduler] Launch Phase")
         with self._monitor:
-            for dev, queue in self._device_queues.items():
-                # Make sure there's an available WorkerThread
+            for dev in self._available_resources.get_resources():
                 if len(self._free_worker_threads) == 0:
                     break
-                if len(queue) > 0: # If there are tasks on the queue.
+                compute_queue = self._compute_task_dev_queues[dev]
+                datamove_queue = self._datamove_task_dev_queues[dev]
+                if len(compute_queue) > 0:
+                    try:
+#print("num of compute tasks on dev:", dev, " :", self._device_compute_task_counts)
+                        if self._device_compute_task_counts[dev] \
+                                < (self._num_colocatable_tasks + 1):
```

Comment:
We'll probably need to revisit this after the deadline I think. 

I think a simple improvement for now might be a separate count for both data move and compute?
Thoughts: @sestephens73 @nicelhc13 ?

Otherwise LGTM. The mutex is a good idea.

Reply:
You may have misread. There currently _is_ a separate count.

Reply:
ah, yep. They just have the same limit. Sorry.